### ISSUE TITLE
GTK+ 3: Toolbar icons too big

### ISSUE THREAD

**Issue by paulmenzel** [geany/geany#2394]
Using Xfce 4.14 on a HiDPI monitor with 2x scaling, using Geany 1.36 with GTK+ 3.24.12 gives a strange toolbar.

![geany-1 36-scale](https://user-images.githubusercontent.com/9318792/68688379-e719f180-056e-11ea-9256-50886b73850d.png)

**Comment by b4n** [comment]
I couldn't reproduce this issue, but I don't have actual HiDPI hardware, and playing with scaling with the GTK debugging tool doesn't expose this behavior.

Useful info I guess:
* This icon is distributed by Geany (*geany-build*)
* The *geany-save-all* icon (4th from the left) we also distribute displays fine
* It is in the *hicolor* theme, whereas the *geany-save-all* one is from the *tango* theme.
* We distribute the same sizes for all our icons, and a scalable version (so it's probably not we don't distribute a needed size)
* The *index.theme* files look very similar (apart from *tango* not providing any *apps* category)

**Comment by elextr** [comment]
> The geany-save-all icon (4th from the left) we also distribute displays fine

Its not displaying the Geany icon, thats why.  Only the fallen over bookcase is a Geany icon.

**Comment by b4n** [comment]
@elextr yes it does, the 4th one is *geany-save-all*, see *icons/tango/scalable/geany-save-all.svg*

**Comment by paulmenzel** [comment]
The Clearlooks style and Xfce Rodent icons are configured under Xfce’s *Appearance* dialog.

The following versions of `geany-save-all` are installed.

    /usr/share/icons/Tango/scalable/actions/geany-save-all.svg
    /usr/share/icons/Tango/48x48/actions/geany-save-all.png
    /usr/share/icons/Tango/32x32/actions/geany-save-all.png
    /usr/share/icons/Tango/24x24/actions/geany-save-all.png
    /usr/share/icons/Tango/16x16/actions/geany-save-all.png
    /usr/share/icons/hicolor/scalable/actions/geany-save-all.svg
    /usr/share/icons/hicolor/48x48/actions/geany-save-all.png
    /usr/share/icons/hicolor/32x32/actions/geany-save-all.png
    /usr/share/icons/hicolor/24x24/actions/geany-save-all.png
    /usr/share/icons/hicolor/16x16/actions/geany-save-all.png

The hicolor folder seems to have more icon sizes.

````
$ ls /usr/share/icons/Tango/
16x16  22x22  24x24  32x32  48x48  icon-theme.cache  index.theme  scalable
$ ls /usr/share/icons/hicolor/
1024x1024  16x16    24x24    36x36    64x64  icon-theme.cache  symbolic
128x128    192x192  256x256  48x48    72x72  index.theme
160x160    22x22    32x32    512x512  96x96  scalable
````

In Geany’s toolbar preferences there is no difference between *System default* and *Large icons*.

With *Small icons* and *Very small icons* it looks like below.

![Screenshot of very small icons](https://user-images.githubusercontent.com/9318792/68760136-7f68b280-0611-11ea-8528-82e0a21fb06d.png)

**Comment by elextr** [comment]
@b4n much as I hate to disagree :) the fourth icon shown in the screenshot above is not the Geany icon, the icon in the screenshot is a box (I guess a disk drive) with two arrows pointing to it, the Geany icon is two floppy disks and a big red X (at least it is in `geany/icons/scalable` on my clone of Geany).

**Comment by b4n** [comment]
@elextr Hence me saying *icons/tango/scalable/geany-save-all.svg*, notice the *tango* in the path ;)

@paulmenzel If there are sizes at which the icon is missing, could you try adding those sizes and see if it helps?
```bash
# To be run from Geany's source directory.
prefix=/usr/ # where Geany is installed
for d in /usr/share/icons/hicolor/[0-9]*x[0-9]*/actions; do
    target=$prefix/${d##/usr/}
    [ -f $d/geany-build.png -o -f $target/geany-build.png ] && continue
    s=${d%x[0-9]*}; s=${s##*/}
    mkdir -p $target
    inkscape -z -e $target/geany-build.png -w $s -h $s icons/scalable/geany-build.svg
done
```
You might also have to update `$prefix/share/icons/hicolor/index.theme`, or not if you installed in `/usr` anyway.

**Comment by elextr** [comment]
@b4n, Maybe I'm being dumb but your original post

> The geany-save-all icon (4th from the left) we also distribute displays fine

But since the icon is not being displayed on the OPs system _how do you know_ it displays fine?

Or do you mean it displays fine on _your_ system which isn't hidpi and so irrelevant?

**Comment by b4n** [comment]
@elextr I'm saying the 4th icon from the left (double arrows pointing to a drive) is a Geany-specific icon, `$srcdir/icons/tango/scalable/geany-save-all.svg` or one of its rendered PNG versions.  And that icon displays fine AFAICT (good size, not blurry).

**Comment by elextr** [comment]
> `$srcdir/icons/tango/scalable/geany-save-all.svg`

Ahh, I been missing the __tango__ in that path.

**Comment by codebrainz** [comment]
Stupid question, and somewhat off-topic, but why is the icon called "geany-save-all" as opposed to "document-save-all"? It seems like the latter has a better chance of getting overridden by the user's icon theme, at least I have several icon themes with "document-save-all" two of which appear to have added symlinks from that to "geany-save-all" because of the naming.

**Comment by codebrainz** [comment]
I have a 3840x2160 display and it looks fine, but I'm on an older version of Xfce which presumably doesn't have scaling (or at least I couldn't find it).

@paulmenzel what happens if you try to temporarily replace the "geany-build" icons with different ones, like "geany-save-all" or whatever other icons? I'm wondering if maybe there's something weird about the SVG code in the "geany-build" icon or something.

**Comment by elextr** [comment]
@codebrainz IIUC XFCE 4.14 is the first release written in GTK3, wouldn't be surprised if it has bugs :)
